### DISCUSSION TITLE
What is the difference between chat-instruct mode and instruct mode?

### DISCUSSION THREAD

**Discussion by majestichou** [oobabooga/text-generation-webui#5417] [Q&A]
In instruct and chat-instruct modes, the instruction template selected under Parameters > Instruction template must match the current model. What is the difference between chat-instruct mode and instruct mode?

**Comment by TechScorpion** [comment]
As far as I know, chat-instruct and instruct modes have different templates for how they will respond (prompt templates?). And I think chat-instruct  still takes stuff from the character profile (which the chat mode does too) but I don't think the instruct mode does. You could check though (see if there's a difference in context size when asking a question or asking it to write something in a new chat). So instruct mode is more general (what they write wouldn't be from a particular character's point of view).

**Comment by araleza** [comment]
There are two types of models available, base models and instruct models.  Base models are trained to _continue what's been written so far_, whereas instruct models are trained to _carry out the instruction / answer the question_.

Base models might seem like a more natural fit for chatting with, especially if they've been trained on a roleplay dataset, as they can see a conversation is taking place and continuing to make conversation is what they'll tend to do automatically.  Instruct models will not tend to understand what to do, because they will be looking for an instruction or question, and that's not what chats looks like.

Chat-instruct mode is used when you want to chat with an instruct trained model.  Behind the scenes, the instruct model is given an instruction to 'write the next reply in the chat', which is an instruction it can follow to effectively be able to carry out a conversation.  You can see that if you do a chat-instruct mode conversation, and then choose 'Send to default' from that little menu box in the corner.

So this Chat-Instruct conversation (with Mixtral-8x7B-instruct):

![image](https://github.com/oobabooga/text-generation-webui/assets/70412719/4ab8f5fd-2665-4368-9b80-c52610e85250)

has the instruction:
```
   'Continue the chat dialogue below. Write a single reply for the character "Chiharu Yamada".'
```
inserted at the top (seen here after 'Send to default' is used), before the character card (which includes an example conversation), and before the actual current conversation:

![image](https://github.com/oobabooga/text-generation-webui/assets/70412719/7124f13d-29f9-45dd-8c7b-6d2f6cc06b78)

If we did the same with Chat mode rather than Chat-Instruct, you can see this in the Default tab instead:

![image](https://github.com/oobabooga/text-generation-webui/assets/70412719/0202091b-93b5-4112-bdc1-3f4c7ba7a542)

Notice that the conversation _still worked_ in chat mode, even though this is an instruct variant of Mixtral, and no instruction was given.  Also, the Mixtral instruct model is looking for [INST] and [/INST] tags around its inputs, and they're just not there in chat mode.  (Other instruct models may look for e.g. '### Instruction:', '### Input:', and '### Response:' markers, known as Alpaca format).  But despite the [INST] and [/INST] tags being missing, it all still happens to work in this case.  That's likely just because models tend to be pretty flexible, as long as there's enough to work out what they should be doing.

One final point to note is that chat-instruct with instruct models **can sometimes get you longer replies**. That's because base models have learned from datasets where people are chatting, and it's rare for people to type more than a sentence or two when chatting.  But instruct models have been fine tuned on datasets where they get a short question, and write multiple paragraphs in reply.  So by using an instruct model, you can often get more lengthy replies, and Chat-Instruct mode can help the model lean in that direction further.

So whether you use a base model in Chat mode or an instruct model in Chat-Instruct mode depends on your personal preference, how the particular model you're using has been trained (i.e. whether it has many chat conversations of the type you like in its training dataset), and how lengthy you want its replies to be.

And just to answer your question, here's Instruct (not Chat-Instruct) mode:

![image](https://github.com/oobabooga/text-generation-webui/assets/70412719/da8b912e-8d17-4bf9-ad0f-254cc1aa1fda)

which leads to this being passed to the model (and including the model's output):

![image](https://github.com/oobabooga/text-generation-webui/assets/70412719/a7406b1e-9567-4419-95ba-4e00d42b2d4c)

@TechScorpion is correct that the prompt template is being used (so Mixtral's [INST] and [/INST] tags are present), and the character profile is not present.

**Reply by majestichou** [reply]
Thanks a lot !! Thank you for your detailed answer. I understand.
